### ISSUE TITLE
[API Proposal]: CollectionsMarshal method enumerating Dictionary<TKey, TValue>'s Values by reference.

### ISSUE THREAD

**Issue by acaly** [dotnet/runtime#58333]
### Background and motivation

`CollectionsMarshal` class has been introduced to expose the collection class's underlying data by reference in order to reduce unnecessary copy of large struct values when accessing only a small part of the value. For `Dictionary<TKey, TValue>`, there are methods that manipulate individual key-value pairs. However, these methods mainly requires accessing from know keys. When the user wants to do batch actions to many elements in the dictionary, enumerating keys and then doing a hash table look up would not be the most efficient way.

Consider a very common use case of `Dictionary<,>` class: the user wants to use it like an in-memory database table, to store large struct values, each containing the data of one row, and all values are indexed by a primary key, which is itself part of the row data but immutable. When reading and writing data from one row with known primary key, existing methods like `GetValueRefOrNullRef` is sufficient. But if the user wants to read all rows, doing some work to each of them (or to some of them that meets specific condition), it would be better if there is a way to enumerate the dictionary returning `ref` to each value.

This issue mainly happens to `Dictionary<,>` because `List<T>` already exposes its `Span<T>`, which can be used to enumerate by a `for (int i = 0; ...` loop to get the references.

### API Proposal

```C#
namespace System.Runtime.InteropServices
{
    public static class CollectionsMarshal
    {
        public static DictionaryValueRefEnumerable<TKey, TValue> EnumerateValuesByRef<TKey, TValue>(Dictionary<TKey, TValue> dictionary) => new(dictionary);

        public struct DictionaryValueRefEnumerable<TKey, TValue>
        {
            internal DictionaryValueRefEnumerable(Dictionary<TKey, TValue> dictionary) { ... }

            DictionaryValueRefEnumerable<TKey, TValue> GetEnumerable() => this;

            public bool MoveNext() { ... }

            public ref TValue Current
            {
                get { ... }
            }
        }
    }
}
```     

### API Usage

```C#
struct Data
{
    public int SomeValue;
    //Other fields.
}
Dictionary<int, Data> dictionary = new();

//...

foreach (ref var data : CollectionsMarshal.EnumerateValuesByRef(dictionary))
{
    if (data.SomeValue != -1)
    {
        data.SomeValue += 1;
    }
}
```     

### Risks

Some design considerations:

* Same safety concerns shared by other `CollectionsMarshal` methods, but this is by design. User should make sure the use is valid, e.g., by guaranteeing no modification is made to the collection during enumerating.
* The returned struct enumerable can only be enumerated directly, and cannot be converted to the `IEnumerable<T>` interface, because the T here is a by-ref type. This should be fine since usually converting to interfaces causes boxing and GC allocation, which is against high-performance requirements.
* Only values can be enumerated, and the key for each item cannot be easily obtained. An alternative way is to return reference to a new struct type of
```c#
public struct ImmutableKeyKeyValuePair<TKey, TValue>
{
    public TKey Key { get; }
    public TValue Value;
}
```
by converting from the corresponding part of the `Dictionary<TKey, TValue>.Entry` struct with `Unsafe.As`.

**Comment by GrabYourPitchforks** [comment]
This would probably benefit from ref fields being a first-class citizen of the runtime \& language (see https://github.com/dotnet/csharplang/issues/1147). It's certainly possible to have an implementation without this, but it's likely to have suboptimal performance.

```cs
public ref struct KeyValuePairByRef<TKey, TValue>
{
    private readonly ref Entry _entry;

    public ref readonly TKey Key => ref _entry.Key;
    public ref TValue Value => ref _entry.Value;
}

public struct TheEnumerator
{
    private readonly Entry[] _entries;
    private int _index;

    public bool MoveNext() { /* checks mimicking current enumerator checks */ }
    public KeyValuePairByRef<TKey, TValue> Current => ref _entries[_index]; // + whatever other checks need take place
}
```

**Comment by acaly** [comment]
> This would probably benefit from ref fields being a first-class citizen of the runtime & language (see [dotnet/csharplang#1147](https://github.com/dotnet/csharplang/issues/1147)). It's certainly possible to have an implementation without this, but it's likely to have suboptimal performance.
> 
> ```cs
> public readonly ref struct KeyValuePairByRef<TKey, TValue>
> {
>     private readonly ref Entry _entry;
> 
>     public ref readonly TKey Key => ref _entry.Key;
>     public ref TValue Value => ref _entry.Value;
> }
> 
> public struct TheEnumerator
> {
>     private readonly Entry[] _entries;
>     private int _index;
> 
>     public bool MoveNext() { /* checks mimicking current enumerator checks */ }
>     public KeyValuePairByRef<TKey, TValue> Current => new (ref _entries[_index]); // + whatever other checks need take place
> }
> ```

I think the difference here is to change one pointer to the node to two pointers to the `TKey` and `TValue` respectively. IMHO Performance-wise these two should be very similar, but the two-pointer methods seems to allow more general implementation of the `Dictionary<,>` class.

**Comment by acaly** [comment]
I have another proposal on the use of `TValue` ref that is not quite related to the proposal but I think might also be good to mention here. Do you think we can remove a key-value pair from the dictionary given the reference to `TValue`? I looked at the current implementation, it seems that's completely possible. The index can be calculated from `Unsafe.ByteOffset`. This might even be possible while enumerating. As long as the dictionary isn't rehashing, the references should remain valid.

**Comment by acaly** [comment]
> This would probably benefit from ref fields being a first-class citizen of the runtime & language (see [dotnet/csharplang#1147](https://github.com/dotnet/csharplang/issues/1147)). It's certainly possible to have an implementation without this, but it's likely to have suboptimal performance.
> 
> ```cs
> public readonly ref struct KeyValuePairByRef<TKey, TValue>
> {
>     private readonly ref Entry _entry;
> 
>     public ref readonly TKey Key => ref _entry.Key;
>     public ref TValue Value => ref _entry.Value;
> }
> 
> public struct TheEnumerator
> {
>     private readonly Entry[] _entries;
>     private int _index;
> 
>     public bool MoveNext() { /* checks mimicking current enumerator checks */ }
>     public KeyValuePairByRef<TKey, TValue> Current => new (ref _entries[_index]); // + whatever other checks need take place
> }
> ```

Since it's inside the corelib, maybe `ByReference<T>` is enough?

**Comment by GrabYourPitchforks** [comment]
> Since it's inside the corelib, maybe `ByReference<T>` is enough?

Perhaps, but we've historically been hesitant to use it for scenarios beyond simply `Span<T>`. It's partly that we don't want to risk propagating a type throughout the runtime when that type has really only been stress tested in one specific scenario, but also that we don't want to make it a habit of relying on runtime tricks for something that should be a first-class feature available to all. (Solving the latter problem would also necessarily imply solving the former, since shipping such a feature would involve properly testing it.)

**Comment by acaly** [comment]
I have update the proposal with some benchmark results. Both methods have some improvements in enumerating (the ratio strongly depends on how fast the hash can be calculated). The second method, when used with removing, is also better than the "standard workaround" of creating another list recording keys, because it's not only faster, but also avoids allocation of the list.

**Comment by GrabYourPitchforks** [comment]
The issue was changed from "allow iterating by references" to "allow removal while iterating", which significantly changes the scope \& direction of this work.

I think you'll have more success proposing an API which tries to split the two concerns. For example, you could consider a `RemoveWhere` method which takes a predicate, which doesn't require any `CollectionsMarshal` or ref field goop. A predicate-based approach would also have a natural "end" state (the `RemoveWhere` method returning), which could allow the dictionary to rebucket its internal entries.

**Comment by stephentoub** [comment]
`Dictionary<>` was updated circa .NET Core 2.1 such that you can remove while iterating; removal does not invalidate enumerators.

**Comment by acaly** [comment]
> `Dictionary<>` was updated circa .NET Core 2.1 such that you can remove while iterating; removal does not invalidate enumerators.

Yes, but it will need another hash table look up, as I mentioned in the background section.

**Comment by stephentoub** [comment]
I'm saying, the background section states "or Dictionary<,>, modifying the content will invalidates the enumerator, and users are forced to temporarily record keys to remove elsewhere, and remove them after enumeration finishes" which is not correct.  This is valid code:
```C#
var d = new Dictionary<int, int> { { 1, 1 }, { 2, 2 }, { 3, 3 } };
foreach (var pair in d)
{
    d.Remove(pair.Key);
}
```

**Comment by acaly** [comment]
> The issue was changed from "allow iterating by references" to "allow removal while iterating", which significantly changes the scope & direction of this work.
> 
> I think you'll have more success proposing an API which tries to split the two concerns. For example, you could consider a `RemoveWhere` method which takes a predicate, which doesn't require any `CollectionsMarshal` or ref field goop. A predicate-based approach would also have a natural "end" state (the `RemoveWhere` method returning), which could allow the dictionary to rebucket its internal entries.

I understand. The reason I think it's probably better to mention it here is because if one method can cover both, it will be easier for user to decide which one to use. If there are two proposals, each add one enumeration method, the API will look much more complcated.

I think this issue might be an initial attempt that illustrates the problem. If you think it's better having an enumerator first, without allowing the removal, I (or anyone else) can open a separate issue to track it. In any case, I still think both proposals are about avoiding repeated hash calculation and look-up while enumerating, and can be seen as one unified problem and solved together.

**Comment by acaly** [comment]
> > as I mentioned in the background section
> 
> I'm saying, the background section states "or Dictionary<,>, modifying the content will invalidates the enumerator, and users are forced to temporarily record keys to remove elsewhere, and remove them after enumeration finishes" which is not correct. This is valid code:
> 
> ```cs
> var d = new Dictionary<int, int> { { 1, 1 }, { 2, 2 }, { 3, 3 } };
> foreach (var pair in d)
> {
>     d.Remove(pair.Key);
> }
> ```

You are right. I will update the wording and probably redo some benchmarks.

**Comment by krwq** [comment]
@acaly can you please update the issue to add example scenarios which this could solve? I personally never needed such functionality but YMMV

**Comment by acaly** [comment]
@krwq Thanks for the reply. I wrote a very brief description on the scenario in the proposal.

> Consider a very common use case of `Dictionary<,>` class: the user wants to use it like an in-memory database table, to store large struct values, each containing the data of one row, and all values are indexed by a primary key, which is itself part of the row data but immutable. When reading and writing data from one row with known primary key, existing methods like `GetValueRefOrNullRef` is sufficient. But if the user wants to read all rows, doing some work to each of them (or to some of them that meets specific condition), it would be better if there is a way to enumerate the dictionary returning `ref` to each value.

I encountered this problem when I was implementing a physics engine in C#. I stored the active interaction data between entities in a dictionary, with `(entity1, entity2, interactionType)` as key and the data (as a value type) as value. This brings some benefits: 1) the collision detection module can use the dictionary to check whether there is already an active interaction between specific entities; 2) using dictionaries avoids unnecessary copy when adding/removing interactions. An alternative approach would be to use reference type instead, but it requires manually pooling instances to make an allocation-free implementation, and it may also sacrifice locality when enumerating interactions.

**Comment by lakecherry** [comment]
I have a similar high performance scenario where the dictionary values (int) all have to be divided by 2 (some kind of decay operation).

How about changing dictionarys' `ValueCollection` enumerator from `public TValue Current => _currentValue!;` to `public ref TValue Current => ref _currentValue!;`

Existing code would still be valid, and ref assignments would give a change option.

**Comment by lakecherry** [comment]
That's true ... but thanks for the quick response!
